File: src/com/easy/RemoveDuplicatesfromSortedList_83.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class RemoveDuplicatesfromSortedList_83 {

    // Iterative version
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 37.8 MB, less than 25.29%
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode mNodeIndex = head;
        while (mNodeIndex.next != null) {
            if (mNodeIndex.val == mNodeIndex.next.val) {
                mNodeIndex.next = mNodeIndex.next.next;
            } else {
                mNodeIndex = mNodeIndex.next;
            }
        }

        return head;
    }

    // Recursive version
    // Runtime: 1 ms, faster than 30.22%
    // Memory Usage: 37.5 MB, less than 66.44%
    public ListNode deleteDuplicates2(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        head.next = deleteDuplicates2(head.next);
        return (head.val == head.next.val)? head.next: head;
    }
}

File: Cpp_src/easy/remove_duplicates_from_sorted_list_83.cc
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// Use iterative
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 12 ms, faster than 74.28%
// Memory Usage: 9.4 MB, less than 39.62%
class Solution
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        if (head == NULL || head->next == NULL)
        {
            return head;
        }
        ListNode *cur = head;

        while (cur->next)
        {
            if (cur->val == cur->next->val)
            {
                ListNode *tmp = cur->next;
                cur->next = cur->next->next;
                delete tmp;
            }
            else
            {
                cur = cur->next;
            }
        }

        return head;
    }
};

// Use recursive
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 12 ms, faster than 74.28%
// Memory Usage: 9.1 MB, less than 100.00%
class Solution2
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        if (head == NULL || head->next == NULL)
        {
            return head;
        }
        head->next = deleteDuplicates(head->next);
        return (head->val == head->next->val) ? head->next : head;
    }
};